tools/mason/MasonRun.chpl

```diff
@@ -24,56 +24,66 @@ use FileSystem;
 use TOML;
 
 proc masonRun(args) {
-  var toParse = open("Mason.toml", iomode.r);
-  var tomlFile = parseToml(toParse);
-  var project = tomlFile["brick"]["name"].s;
-  var show = false;
-  var execopts: [1..0] string;
-  if args.size > 2 {
-    for arg in args[2..] {
-      if arg == '-h' || arg == '--help' {
-        masonRunHelp();
-        exit();
-      }
-      else if arg == '--build' {
-        masonBuild(['mason', 'build']);
-      }
-      else if arg == '--show' {
-        show = true;
-      }
-      else if arg == 'run' {
-        continue;
-      }
-      else {
-        execopts.push_back(arg);
+
+  try! {
+
+    const cwd = getEnv("PWD");
+    const projectHome = getProjectHome(cwd);
+    const toParse = open(projectHome + "/Mason.toml", iomode.r);
+    const tomlFile = parseToml(toParse);
+    const project = tomlFile["brick"]["name"].s;
+    var show = false;
+    var execopts: [1..0] string;
+
+    if args.size > 2 {
+      for arg in args[2..] {
+        if arg == '-h' || arg == '--help' {
+          masonRunHelp();
+          exit();
+        }
+        else if arg == '--build' {
+          masonBuild(['mason', 'build']);
+        }
+        else if arg == '--show' {
+          show = true;
+       }
+       else if arg == 'run' {
+          continue;
+       }
+       else {
+          execopts.push_back(arg);
+       }
       }
     }
-  }
-  // Find the Binary and execute
-  if isDir('target') {
-    var execs = ' '.join(execopts);
-    var command = "target/debug/" + project + ' ' + execs;
-    if isDir('target/release') then
-      command = "target/release/" + project + ' ' + execs;
+    // Find the Binary and execute
+    if isDir(projectHome + '/target') {
+      var execs = ' '.join(execopts);
+      var command = projectHome + "/target/debug/" + project + ' ' + execs;
+      if isDir(projectHome + '/target/release') then
+        command = projectHome + "/target/release/" + project + ' ' + execs;
 
-    if show then
-      writeln("Executing binary: " + command);
+      if show then
+        writeln("Executing binary: " + command);
 
-    if isFile("Mason.lock") then  // If built
-      runCommand(command);
-    else if isFile("Mason.toml") { // If not built
-      masonBuild(args);
-      runCommand(command);
-    }
-    else writeln("call mason run from the top level of your projects directory");
+      if isFile(projectHome + "/Mason.lock") then  // If built
+        runCommand(command);
+      else if isFile(projectHome + "/Mason.toml") { // If not built
+        masonBuild(args);
+        runCommand(command);
+      }
+      else writeln("Mason could not find your Mason.lock file");
 
-    // Close memory
-    delete tomlFile;
-    toParse.close();
+      // Close memory
+      delete tomlFile;
```

Comment:
Bonus points if you apply the new delete-free memory management here. 

I believe that would involve changing the above assignment to:

```chpl
const tomlFile = new Owned(parseToml(toParse));
```

and then you can remove the `delete`here.

(@mppf can spot check me on that)